### PR TITLE
[JENKINS-41243] Slightly faster tagging of synthetic stages

### PR DESCRIPTION
**Pull request by abayer**
[JENKINS-41243](https://issues.jenkins-ci.org/browse/JENKINS-41243)

This approach yielded an average time from just before the stage call
to just before actually attaching the tag of 15.4ms over 5 runs, vs
40.8ms over 5 runs with the previous approach. I'm sure there are many
things wrong with this approach, though.

cc @reviewbybees esp @rsandell @svanoort @vivek @jglick

### CODE REVIEW

**Comment by abayer** [comment]
@jglick Well, `StageStep` is final, so I couldn't do an extension of it, and I assumed that it'd be a Bad Thing to reimplement all of `StageStep` with some sugar for this.

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/SyntheticStageGraphListener.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                }
+                if (SyntheticStageNames.postStages().contains(node.getDisplayName())) {
+                    attachTag(node, SyntheticStage.getPost());
+                }
+            }
+        }
+    }
+
+    private void attachTag(FlowNode currentNode, String syntheticContext) {
+        TagsAction tagsAction = currentNode.getAction(TagsAction.class);
```
Comment (svanoort) [inline pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/SyntheticStageGraphListener.java]: Suggest using getPersistentAction instead (best practice, it's a bit faster and lighter).

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/SyntheticStageGraphListener.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        if (tagsAction == null) {
+            tagsAction = new TagsAction();
+            tagsAction.addTag(SyntheticStage.TAG_NAME, syntheticContext);
+            currentNode.addAction(tagsAction);
+        } else if (tagsAction.getTagValue(SyntheticStage.TAG_NAME) == null) {
+            tagsAction.addTag(SyntheticStage.TAG_NAME, syntheticContext);
+            try {
+                currentNode.save();
+            } catch (IOException e) {
+                // TODO: should probably do something here? Would prefer to be calling FlowNode.persistSafe() but it's private
```
Comment (svanoort) [inline pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/SyntheticStageGraphListener.java]: 🐛  You need to handle it somehow -- at least log it.

Comment (abayer) [inline pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/SyntheticStageGraphListener.java]: Hence the `TODO`. =) I'll just ape `FlowNode.persistSafe()`, I guess.

**Comment by abayer** [comment]
@svanoort Yeah, very much hasn't touched on the resume issue. I wanted to try to make sure it was even worth pursuing before going that far.

**Comment by abayer** [comment]
@svanoort Made some changes - thoughts welcome/desired. =)

**Comment by jglick** [comment]
> a Bad Thing to reimplement all of `StageStep`

You mean this?

```java
getContext().newBodyInvoker().withCallback(BodyExecutionCallback.wrap(getContext())).withDisplayName(step.name).start();
return false;
```

**Comment by abayer** [comment]
@jglick Plus `StageAction`, plus getting Blue Ocean and Stage View to grok it, etc...

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy [hunk 3]
```diff
@@ -193,21 +190,15 @@ public class Utils {
...
         stages.removeSourceLocation()
 
-        CpsThread thread = CpsThread.current()
-        CpsFlowExecution execution = thread?.execution
-
-        if (execution != null && !execution.complete) {
-            execution.addListener(new SyntheticStageGraphListener())
-        }
-
+        r.addAction(new SyntheticStageGraphListener.GraphListenerAction())
```
Comment (svanoort) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy]: Check if action exists before adding it.

Comment (svanoort) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy]: 🐛

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/SyntheticStageGraphListener.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+import java.io.IOException;
+import java.util.logging.Logger;
+
+import static java.util.logging.Level.WARNING;
+
+public final class SyntheticStageGraphListener implements GraphListener {
+    private static final Logger LOGGER = Logger.getLogger(SyntheticStageGraphListener.class.getName());
+
+    @Override
+    public void onNewHead(FlowNode node) {
```
Comment (svanoort) [inline pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/SyntheticStageGraphListener.java]: @abayer Not a huge fan of using the displayName here -- can we do getPersistentAction(LabelAction.class) instead -- there are solid performance reasons why you don't want to do the displayName (much more complex).

Comment (abayer) [inline pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/SyntheticStageGraphListener.java]: Sure thing.

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/SyntheticStageGraphListener.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                if (SyntheticStageNames.postStages().contains(node.getDisplayName())) {
+                    attachTag(node, SyntheticStage.getPost());
+                }
+            }
+        }
+    }
+
+    private void attachTag(FlowNode currentNode, String syntheticContext) {
+        TagsAction tagsAction = currentNode.getPersistentAction(TagsAction.class);
+        if (tagsAction == null) {
```
Comment (svanoort) [inline pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/SyntheticStageGraphListener.java]: I had an API that would do this in TagsAction at one point... removed it during review IIRC, but maybe worth reconsidering

**Comment by abayer** [comment]
Merging once tests pass!

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/SyntheticStageGraphListener.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            }
+        }
+    }
+
+    public static class GraphListenerAction extends InvisibleAction implements RunAction2 {
+        @Override
+        public void onLoad(Run<?, ?> r) {
+            if (r != null && r instanceof WorkflowRun) {
+                WorkflowRun run = (WorkflowRun) r;
+                attachListener(run);
```
Comment (jglick) [inline pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/SyntheticStageGraphListener.java]: [JENKINS-43055](https://issues.jenkins-ci.org/browse/JENKINS-43055)

**Pull request merged**